kep/206-resource-limit-aware-scoring/README.md

```diff
@@ -0,0 +1,96 @@
+  - [Implementation History](#implementation-history)
+  - [References](#references)
+<!-- /toc -->
+
+## Motivation
+
+Burstable pods in Kubernetes can have higher pod resource limits than resource requests. Because Kubernetes scheduler only takes into account resource requests during pod scheduling, resource limits on a node can be over-subscribed, i.e., the total resource limits of scheduled pods’ on a node exceeding the node’s allocatable capacity. Burstable pods typically have different limit to request ratios. As a result, resource limit to allocatable ratios on different nodes in a cluster (i.e., the total resource limit of running pods on a node to the node’s allocatable ratio) can vary a lot. We have noticed nodes' limit to allocatable ratio vary from 0.1 to 6 in production clusters. over-subscription can cause resource contention, performance issues and even pod failure (e.g., OOM). 
+
+To reduce the extent of resource over-subscription and associated resource contention risks, we propose a scoring plugin that considers a node's resource limit in scheduling.
+
+## Goals
+Develop and add a scoring plugin to mitigate resource over-subscription issue caused by burstable pods through "spreading" or "balancing" pod's resource limits across nodes. 
+
+## Non-Goals
+Replace existing resource allocated plugin. 
+
+## Use Cases
+Consider two nodes with the same resource allocatable (8 CPU cores) and each has two pods running.
+- node1: pod1(request: 2, limit: 6), pod2(request: 2, limit: 4); total request/allocatable: 4/8, total limit/allocatable: 10/8
+- node2: pod3(request: 3, limit: 3), pod4(request: 2, limit: 2); total request/allocatable: 5/8, total limit/allocatable: 5/8
+
+**Default scheduler**
+
+To schedule a new pod pod5 (request: 1, limit: 4), the default scheduler with `LeastAllocated` will choose node1 over node2 as node1 has a smaller request/allocatable ratio (4/8 vs. 5/8). The resulting node resource status will be
+- node1: request/allocatable = 5/8, limit/allocatable = 14/8  
+- node2: request/allocatable = 5/8, limit/allocatable = 5/8
+
+node1's resource limit is oversubscribed with a limit to allocatable ratio 14/8.  It will have a high risk of resource contention.
+
+**Proposed scheduler**
+
+If we consider node1's resource limit is already oversubscribed and hence place pod5 on node 2 instead, the result will be 
+- node1: total request/allocatable = 4/8, total limit/allocatable = 10/8  
+- node2: total request/allocatable = 6/8, total limit/allocatable = 9/8
+
+node1’s total resource limit remains the same and node2 has a total resource limit of 9. node2 will have a lower risk of resource contention with pod5 running on it. 
+
+
+## Terms
+
+## Proposal
+
+A custom scheduler scoring plugin scores a node by considering existing pods’ resource limits to the node's allocatable ratio on the node. It attempts to “evenly” spread or distribute pod resource limits across nodes in a cluster. 
+
+## Design Details
```

Comment:
One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

Reply:
> One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

Added an implementation section describing the details.  `limits` can be calculated from cached `NodeInfo.Pods (pod.Spec.Containers)`.

Reply:
> One piece missing in this section (or can be added into the "Implementation" section): how would you like to cache the `limits` of each node? It's not provided by scheduler framework's Snapshot - which has Requested and Allocatable only. So I'd expect to see the details explaining how the `limits` are managed.

It'll be calculated from cached `NodeInfo.Pods`. Add the detail to the proposal. 

   for _, pod := range nodeInfo.Pods {
                for _, container := range pod.Pod.Spec.Containers {
                       ...
                }
  }